File: services/tradelog/tradelog-api/src/test/java/com/example/tradelog/api/validator/ValidateShareJournalModelTest.java
package com.example.tradelog.api.validator;

import com.example.tradelog.api.spec.model.ActionType;
import com.example.tradelog.api.spec.model.ShareJournalModel;
import com.example.tradelog.api.spec.model.TransactionModel;
import com.example.tradelog.api.spec.model.TransactionSettingsModel;
import com.example.tradelog.api.spec.model.TransactionType;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.time.OffsetDateTime;

class ValidateShareJournalModelTest {

    @Test
    void testValidateShareJournalModelWhenNull() {
        Assertions.assertFalse(RequestValidation.validateShareJournalModel.test(null));
    }


    @Test
    void testValidateShareJournalModelWhenActionIsInvalid() {
        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
        TransactionModel transactionModel = new TransactionModel("1234", "123", OffsetDateTime.now(), "XYZ", TransactionType.SHARE, 0, transactionSettingsModel);
        ShareJournalModel model = new ShareJournalModel(transactionModel, 1, 1, 1, ActionType.UNKNOWN);

        Assertions.assertFalse(RequestValidation.validateShareJournalModel.test(model));
    }

    @Test
    void testValidateShareJournalModelWhenAccountIsEmtpy() {
        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
        TransactionModel transactionModel = new TransactionModel("1234", "", OffsetDateTime.now(), "XYZ", TransactionType.SHARE, 0, transactionSettingsModel);
        ShareJournalModel model = new ShareJournalModel(transactionModel, 1, 1, 1, ActionType.BUY);

        Assertions.assertFalse(RequestValidation.validateShareJournalModel.test(model));
    }

    @Test
    void testValidateShareJournalModelWhenPriceIsNegative() {
        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
        TransactionModel transactionModel = new TransactionModel("1234", "123", OffsetDateTime.now(), "XYZ", TransactionType.SHARE, 0, transactionSettingsModel);
        ShareJournalModel model = new ShareJournalModel(transactionModel, -1, 1, 1, ActionType.BUY);

        Assertions.assertFalse(RequestValidation.validateShareJournalModel.test(model));
    }

    @Test
    void testValidateShareJournalModelWhenQuantityIsZero() {
        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
        TransactionModel transactionModel = new TransactionModel("1234", "123", OffsetDateTime.now(), "XYZ", TransactionType.SHARE, 0, transactionSettingsModel);
        ShareJournalModel model = new ShareJournalModel(transactionModel, 1, 1, 0, ActionType.BUY);

        Assertions.assertFalse(RequestValidation.validateShareJournalModel.test(model));
    }

    @Test
    void testValidateShareJournalModelWhenSymbolIsEmpty() {
        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
        TransactionModel transactionModel = new TransactionModel("1234", "123", OffsetDateTime.now(), "", TransactionType.SHARE, 0, transactionSettingsModel);
        ShareJournalModel model = new ShareJournalModel(transactionModel, 1, 1, 1, ActionType.BUY);

        Assertions.assertFalse(RequestValidation.validateShareJournalModel.test(model));
    }

    @Test
    void testValidateShareJournalModelWhenSymbolIsSpace() {
        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
        TransactionModel transactionModel = new TransactionModel("1234", "123", OffsetDateTime.now(), " ", TransactionType.SHARE, 0, transactionSettingsModel);
        ShareJournalModel model = new ShareJournalModel(transactionModel, 1, 1, 1, ActionType.BUY);

        Assertions.assertFalse(RequestValidation.validateShareJournalModel.test(model));
    }

    @Test
    void testValidateShareJournalModelWhenTransactionTypeIsNotStock() {
        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
        TransactionModel transactionModel = new TransactionModel("1234", "123", OffsetDateTime.now(), "XYZ", TransactionType.UNKNOWN, 0, transactionSettingsModel);
        ShareJournalModel model = new ShareJournalModel(transactionModel, 1, 1, 1, ActionType.BUY);

        Assertions.assertFalse(RequestValidation.validateShareJournalModel.test(model));
    }

    @Test
    void testValidateShareJournalModelTruthy() {
        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
        TransactionModel transactionModel = new TransactionModel("1234", "123456789012345678901234567890123456", OffsetDateTime.now(), "XYZ", TransactionType.SHARE, 0, transactionSettingsModel);
        ShareJournalModel model = new ShareJournalModel(transactionModel, 10.00, 1, 100, ActionType.BUY);

        Assertions.assertTrue(RequestValidation.validateShareJournalModel.test(model));
    }
}

File: services/tradelog/tradelog-api/src/test/kotlin/com/example/tradelog/api/rest/validator/ValidateShareJournalModelTest.kt
package com.example.tradelog.api.rest.validator

import com.example.tradelog.api.spec.model.TLShareJournalDto
import com.example.tradelog.api.spec.model.TLTransactionDto
import com.example.tradelog.api.spec.model.TLTransactionSettingsDto
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.time.OffsetDateTime

class ValidateShareJournalModelTest {

    @Test
    fun testValidateShareJournalModelWhenActionIsInvalid() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setAccountId("123")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.SHARE)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val shareJournalDto = TLShareJournalDto.newBuilder()
                .setAction(TLShareJournalDto.ActionType.UNKNOWN_ACTION_TYPE)
                .setPrice(1.0)
                .setQuantity(1)
                .setTransactionDetails(transactionDto)
                .build()

        Assertions.assertFalse(RequestValidator.validateShareJournalModel(shareJournalDto))
    }


    @Test
    fun testValidateShareJournalModelWhenAccountIsEmpty() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.SHARE)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val shareJournalDto = TLShareJournalDto.newBuilder()
                .setAction(TLShareJournalDto.ActionType.BUY)
                .setPrice(1.0)
                .setQuantity(1)
                .setTransactionDetails(transactionDto)
                .build()

        Assertions.assertFalse(RequestValidator.validateShareJournalModel(shareJournalDto))
    }


    @Test
    fun testValidateShareJournalModelWhenPriceIsNegative() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.SHARE)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val shareJournalDto = TLShareJournalDto.newBuilder()
                .setAction(TLShareJournalDto.ActionType.BUY)
                .setPrice(-11.0)
                .setQuantity(1)
                .setTransactionDetails(transactionDto)
                .build()

        Assertions.assertFalse(RequestValidator.validateShareJournalModel(shareJournalDto))
    }


    @Test
    fun testValidateShareJournalModelWhenQuantityIsZero() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.SHARE)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val shareJournalDto = TLShareJournalDto.newBuilder()
                .setAction(TLShareJournalDto.ActionType.BUY)
                .setPrice(1.0)
                .setQuantity(-1)
                .setTransactionDetails(transactionDto)
                .build()

        Assertions.assertFalse(RequestValidator.validateShareJournalModel(shareJournalDto))
    }


    @Test
    fun testValidateShareJournalModelWhenSymbolIsEmpty() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("")
                .setType(TLTransactionDto.TransactionType.SHARE)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val shareJournalDto = TLShareJournalDto.newBuilder()
                .setAction(TLShareJournalDto.ActionType.BUY)
                .setPrice(1.0)
                .setQuantity(1)
                .setTransactionDetails(transactionDto)
                .build()

        Assertions.assertFalse(RequestValidator.validateShareJournalModel(shareJournalDto))
    }


    @Test
    fun testValidateShareJournalModelWhenSymbolIsSpace() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol(" ")
                .setType(TLTransactionDto.TransactionType.SHARE)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val shareJournalDto = TLShareJournalDto.newBuilder()
                .setAction(TLShareJournalDto.ActionType.BUY)
                .setPrice(1.0)
                .setQuantity(1)
                .setTransactionDetails(transactionDto)
                .build()

        Assertions.assertFalse(RequestValidator.validateShareJournalModel(shareJournalDto))
    }


    @Test
    fun testValidateShareJournalModelWhenTransactionTypeIsNotStock() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.OPTION)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val shareJournalDto = TLShareJournalDto.newBuilder()
                .setAction(TLShareJournalDto.ActionType.BUY)
                .setPrice(1.0)
                .setQuantity(1)
                .setTransactionDetails(transactionDto)
                .build()

        Assertions.assertFalse(RequestValidator.validateShareJournalModel(shareJournalDto))
    }


    @Test
    fun testValidateShareJournalModelTruthy() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setAccountId("123456789012345678901234567890123456")
                .setDate(OffsetDateTime.now().toString())
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.SHARE)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val shareJournalDto = TLShareJournalDto.newBuilder()
                .setAction(TLShareJournalDto.ActionType.BUY)
                .setPrice(1.0)
                .setQuantity(1)
                .setTransactionDetails(transactionDto)
                .build()

        Assertions.assertTrue(RequestValidator.validateShareJournalModel(shareJournalDto))
    }
}